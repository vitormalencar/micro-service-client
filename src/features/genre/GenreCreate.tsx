import { Box, Paper, Typography } from "@mui/material";
import { useSnackbar } from "notistack";
import { useEffect, useState } from "react";
import { Genre } from "../../types/Genres";
import { GenreForm } from "./components/GenreFrom";
import {
  initialState as genreInintalState,
  useCreateGenreMutation,
  useGetCaTegoriesQuery,
} from "./genreSlice";
import { mapGenreToForm } from "./util";

export const GenreCreate = () => {
  const { enqueueSnackbar } = useSnackbar();
  const { data: categories } = useGetCaTegoriesQuery();
  const [createGenre, status] = useCreateGenreMutation();
  const [genreState, setGenreState] = useState<Genre>(genreInintalState);

  function handleChange(e: React.ChangeEvent<HTMLInputElement>) {
    const { name, value } = e.target;
    setGenreState({ ...genreState, [name]: value });
  }

  async function handleSubmit(e: React.FormEvent<HTMLFormElement>) {
    e.preventDefault();
    await createGenre(mapGenreToForm(genreState));
  }

  useEffect(() => {
    if (status.isSuccess) {
      enqueueSnackbar(`Genre created`, { variant: "success" });
    }
    if (status.isError) {
      enqueueSnackbar(`Genre not created`, { variant: "error" });
    }
  }, [status, enqueueSnackbar]);

  return (
    <Box>
      <Paper>
        <Box p={2}>
          <Box mb={2}>
            <Typography variant="h4">Create Genre</Typography>
          </Box>
        </Box>

        <GenreForm
          genre={genreState}
          categories={categories?.data}
          isLoading={status.isLoading}
          isDisabled={status.isLoading}
          handleSubmit={handleSubmit}
          handleChange={handleChange}
        />
      </Paper>
    </Box>
  );
};
