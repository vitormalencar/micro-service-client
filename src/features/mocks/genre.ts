export const genreResponse = {
  data: [
    {
      id: "1",
      name: "Norfolk Island",
      is_active: true,
      deleted_at: null,
      created_at: "2022-10-20T08:28:28+0000",
      updated_at: "2022-10-20T08:28:28+0000",
      categories: [
        {
          id: "ba994b75-2b8d-4773-abe2-6b43424a2677",
          name: "PaleTurquoise",
          description: "Quae quo pariatur ut doloribus consequatur.",
          is_active: true,
          deleted_at: null,
          created_at: "2022-10-20T08:28:21+0000",
          updated_at: "2022-10-20T08:28:21+0000",
          pivot: {
            genre_id: "1",
            category_id: "ba994b75-2b8d-4773-abe2-6b43424a2677",
          },
        },
      ],
    },
  ],
  links: {
    first: "http://localhost:8000/api/genres?page=1",
    last: "http://localhost:8000/api/genres?page=7",
    prev: null,
    next: "http://localhost:8000/api/genres?page=2",
  },
  meta: {
    current_page: 1,
    from: 1,
    last_page: 7,
    path: "http://localhost:8000/api/genres",
    per_page: 15,
    to: 15,
    total: 99,
  },
};

export const genreResponsePage2 = {
  data: [
    {
      id: "2",
      name: "Norfolk Island 2",
      is_active: true,
      deleted_at: null,
      created_at: "2022-10-20T08:28:28+0000",
      updated_at: "2022-10-20T08:28:28+0000",
      categories: [
        {
          id: "ba994b75-2b8d-4773-abe2-6b43424a2677",
          name: "PaleTurquoise",
          description: "Quae quo pariatur ut doloribus consequatur.",
          is_active: true,
          deleted_at: null,
          created_at: "2022-10-20T08:28:21+0000",
          updated_at: "2022-10-20T08:28:21+0000",
          pivot: {
            genre_id: "2",
            category_id: "ba994b75-2b8d-4773-abe2-6b43424a2677",
          },
        },
      ],
    },
  ],
  links: {
    first: "http://localhost:8000/api/genres?page=1",
    last: "http://localhost:8000/api/genres?page=7",
    prev: "http://localhost:8000/api/genres?page=1",
    next: "http://localhost:8000/api/genres?page=3",
  },
  meta: {
    current_page: 2,
    from: 16,
    last_page: 7,
    path: "http://localhost:8000/api/genres",
    per_page: 15,
    to: 30,
    total: 99,
  },
};
