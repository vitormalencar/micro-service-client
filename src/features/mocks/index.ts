export const categoryResponse = {
  data: [
    {
      id: "cbdd550c-ad46-4e50-be8d-a8266aff4162",
      name: "PaleTurquoise",
      description: "Explicabo nemo voluptate aut nostrum impedit minus.",
      is_active: true,
      deleted_at: null,
      created_at: "2022-09-27T17:10:33+0000",
      updated_at: "2022-09-27T17:10:33+0000",
    },
    {
      id: "c9f5b9b9-9b9a-4b9a-8b9a-9b9a9b9a9b9a",
      name: "PapayaWhip",
      description: "Quia voluptatem voluptatem.",
      is_active: true,
      deleted_at: null,
      created_at: "2022-09-27T17:10:33+0000",
      updated_at: "2022-09-27T17:10:33+0000",
    },
  ],
  links: {
    first: "http://localhost:8000/api/categories?page=1",
    last: "http://localhost:8000/api/categories?page=7",
    prev: null,
    next: "http://localhost:8000/api/categories?page=2",
  },
  meta: {
    current_page: 1,
    from: 1,
    last_page: 7,
    path: "http://localhost:8000/api/categories",
    per_page: 15,
    to: 15,
    total: 95,
  },
};

export const categoryResponsePage2 = {
  data: [
    {
      id: "cbdd550c-ad46-4e50-be8d-a8266aff4163",
      name: "SeaGreen",
      description: "Explicabo nemo voluptate aut nostrum impedit minus.",
      is_active: true,
      deleted_at: null,
      created_at: "2022-09-27T17:10:33+0000",
      updated_at: "2022-09-27T17:10:33+0000",
    },
  ],
  links: {
    first: "http://localhost:8000/api/categories?page=1",
    last: "http://localhost:8000/api/categories?page=7",
    prev: "http://localhost:8000/api/categories?page=1",
    next: "http://localhost:8000/api/categories?page=3",
  },
  meta: {
    current_page: 2,
    from: 1,
    last_page: 7,
    path: "http://localhost:8000/api/categories",
    per_page: 15,
    to: 15,
    total: 95,
  },
};

export const castMemberResponse = {
  data: [
    {
      id: "f55fca48-d422-48bf-b212-956215eddcaf",
      name: "Teste",
      type: 1,
      deleted_at: null,
      created_at: "2022-10-03T16:23:27+0000",
      updated_at: "2022-10-03T16:23:27+0000",
    },
  ],
  links: {
    first: "http://localhost:8000/api/cast_members?page=1",
    last: "http://localhost:8000/api/cast_members?page=7",
    prev: null,
    next: "http://localhost:8000/api/cast_members?page=2",
  },
  meta: {
    current_page: 1,
    from: 1,
    last_page: 7,
    path: "http://localhost:8000/api/cast_members",
    per_page: 15,
    to: 15,
    total: 100,
  },
};

export const castMemberResponsePage2 = {
  data: [
    {
      id: "f55fca48-d422-48bf-b212-956215eddcae",
      name: "Teste 2",
      type: 1,
      deleted_at: null,
      created_at: "2022-10-03T16:23:27+0000",
      updated_at: "2022-10-03T16:23:27+0000",
    },
  ],
  links: {
    first: "http://localhost:8000/api/cast_members?page=1",
    last: "http://localhost:8000/api/cast_members?page=7",
    prev: "http://localhost:8000/api/cast_members?page=1",
    next: "http://localhost:8000/api/cast_members?page=3",
  },
  meta: {
    current_page: 2,
    from: 1,
    last_page: 7,
    path: "http://localhost:8000/api/cast_members",
    per_page: 15,
    to: 15,
    total: 100,
  },
};
