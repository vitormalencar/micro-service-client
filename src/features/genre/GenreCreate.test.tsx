import { rest } from "msw";
import { setupServer } from "msw/node";
import {
  fireEvent,
  renderWithProviders,
  screen,
  waitFor,
} from "../../utils/test-utils";
import { GenreCreate } from "./GenreCreate";
import { baseUrl } from "../api/apiSlice";
import { categoryResponse } from "../mocks";

const handlers = [
  rest.get(`${baseUrl}/categories`, (req, res, ctx) => {
    return res(ctx.json(categoryResponse));
  }),

  rest.post(`${baseUrl}/genres`, (_, res, ctx) => {
    return res(ctx.status(201));
  }),
];

const server = setupServer(...handlers);

describe("GenreCreate", () => {
  afterAll(() => server.close());
  beforeAll(() => server.listen());
  afterEach(() => server.resetHandlers());

  it("should render correctly", () => {
    const { asFragment } = renderWithProviders(<GenreCreate />);
    expect(asFragment()).toMatchSnapshot();
  });

  it("should handle submit", async () => {
    renderWithProviders(<GenreCreate />);
    const name = screen.getByTestId("name");
    const submit = screen.getByText("Save");

    await waitFor(() => {
      expect(submit).toBeInTheDocument();
    });

    fireEvent.change(name, { target: { value: "test" } });
    fireEvent.click(submit);

    await waitFor(() => {
      const text = screen.getByText("Genre created");
      expect(text).toBeInTheDocument();
    });
  });

  it("should handle error", async () => {
    server.use(
      rest.post(`${baseUrl}/genres`, (_, res, ctx) => {
        return res(ctx.status(500));
      })
    );

    renderWithProviders(<GenreCreate />);
    const name = screen.getByTestId("name");
    const submit = screen.getByText("Save");

    await waitFor(() => {
      expect(submit).toBeInTheDocument();
    });

    fireEvent.change(name, { target: { value: "test" } });
    fireEvent.click(submit);

    await waitFor(() => {
      const text = screen.getByText("Genre not created");
      expect(text).toBeInTheDocument();
    });
  });
});
