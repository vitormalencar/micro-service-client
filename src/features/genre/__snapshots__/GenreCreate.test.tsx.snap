// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenreCreate should render correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-0"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1ps6pg7-MuiPaper-root"
    >
      <div
        class="MuiBox-root css-19midj6"
      >
        <div
          class="MuiBox-root css-1qm1lh"
        >
          <h4
            class="MuiTypography-root MuiTypography-h4 css-5lbw0b-MuiTypography-root"
          >
            Create Genre
          </h4>
        </div>
      </div>
      <div
        class="MuiBox-root css-19midj6"
      >
        <form>
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 css-zow5z4-MuiGrid-root"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
            >
              <div
                class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
              >
                <div
                  class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
                >
                  <label
                    class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                    data-shrink="false"
                    for=":r0:"
                    id=":r0:-label"
                  >
                    Name
                    <span
                      aria-hidden="true"
                      class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                    >
                       *
                    </span>
                  </label>
                  <div
                    class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
                  >
                    <input
                      aria-invalid="false"
                      class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                      data-testid="name"
                      id=":r0:"
                      name="name"
                      required=""
                      type="text"
                      value=""
                    />
                    <fieldset
                      aria-hidden="true"
                      class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                    >
                      <legend
                        class="css-1ftyaf0"
                      >
                        <span>
                          Name *
                        </span>
                      </legend>
                    </fieldset>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
            >
              <div
                class="MuiAutocomplete-root MuiAutocomplete-hasPopupIcon css-16awh2u-MuiAutocomplete-root"
                data-testid="categories-search"
              >
                <div
                  class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
                  data-testid="categories-input"
                >
                  <label
                    class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-disabled css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                    data-shrink="false"
                    for=":r1:"
                    id=":r1:-label"
                  >
                    Categories
                  </label>
                  <div
                    class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-disabled MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
                  >
                    <input
                      aria-autocomplete="list"
                      aria-expanded="false"
                      aria-invalid="false"
                      autocapitalize="none"
                      autocomplete="off"
                      class="MuiInputBase-input MuiOutlinedInput-input Mui-disabled MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
                      disabled=""
                      id=":r1:"
                      role="combobox"
                      spellcheck="false"
                      type="text"
                      value=""
                    />
                    <div
                      class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
                    >
                      <button
                        aria-label="Open"
                        class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
                        disabled=""
                        tabindex="-1"
                        title="Open"
                        type="button"
                      >
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                          data-testid="ArrowDropDownIcon"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M7 10l5 5 5-5z"
                          />
                        </svg>
                      </button>
                    </div>
                    <fieldset
                      aria-hidden="true"
                      class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                    >
                      <legend
                        class="css-1ftyaf0"
                      >
                        <span>
                          Categories
                        </span>
                      </legend>
                    </fieldset>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
            >
              <div
                class="MuiBox-root css-10egq61"
              >
                <a
                  class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
                  href="/genres"
                  tabindex="0"
                >
                  Back
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </a>
                <button
                  class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium css-zcbmsk-MuiButtonBase-root-MuiButton-root"
                  tabindex="0"
                  type="submit"
                >
                  Save
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </div>
            </div>
          </div>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
